### DISCUSSION TITLE
Dose iModeljs support WASD key control under the Walk Around?

### DISCUSSION THREAD

**Discussion by Simplefair** [iTwin/itwinjs-core#721] [Q&A] [answered]
When I set the iModel display environment to "Walk Around", I have to control the walk direction by mouse. Does iModeljs support WASD key control like playing a game?

**Comment by pmconne** [comment, marked as answer]
Yes - the [LookAndMoveTool](https://github.com/imodeljs/imodeljs/blob/98f07cc1e7a677b54252c5874521038ca5708281/core/frontend/src/tools/ViewTool.ts#L3011).
Mouse to look around. WASD or arrow keys to walk. Q and E to move up/down. Mouse wheel or +/- to change speed. 
It currently supports only rudimentary collision detection based on what you're looking at. C to toggle collision detection with walls. Z to toggle collision detection with floor.

**Comment by Simplefair** [comment]
Thank  you very much

**Comment by hanan-commits-work** [comment]
Just a followup for this feature, it it possible that this feature remains activated throughout the runtime of the app, instead of switching it on and off like a toggle (just like a browser, we may use arrow keys to scroll or we may use mouse wheel), if so, then what is the approach to this?

**Comment by pmconne** [comment]
> instead of switching it on and off like a toggle

Do you mean, needing to click in the view to activate look-and-move mode?
If so, not really. The tool uses [pointer events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events) to take control of the mouse cursor, which prevents the user from interacting with the rest of the app's UI.
If not, please clarify.

**Reply by hanan-commits-work** [reply]
Hi, Let me elaborate on what we are trying to do.

Basically, we want users to be able to perform navigation however they want. Like users should be able to move around using keyboard and mouse. 
Currently, the users have to click on a button to activate the keyboard navigation and then move around using WASD keys. But we want to skip this entirely and make users to able to use keys without clicking on a button. So the users can use WASD keys to move around throughout the runtime.

Hope that makes sense, if not then please let me know.

**Reply by pmconne** [reply]
The keys are not the problem - that's the "move" part of the LookAndMoveTool. You could easily modify the tool to permit movement without requiring a button click to activate it.
The problem is the mouse - the "look" part. The tool [takes full control of the mouse](https://developer.mozilla.org/en-US/docs/Web/API/Element/requestPointerLock) so it becomes impossible for the user to click on anything else in viewport.
You could modify the tool to use the keyboard instead of mouse to look around.
Or, if your UI contains nothing else for the user to click on, you could try to modify the tool to permit mouse look without requiring a click. I say "try to" because conformant web browsers [explicitly require user interaction](https://developer.mozilla.org/en-US/docs/Web/Security/User_activation) to enable pointer lock.

**Reply by bbastings** [reply]
> Currently, the users have to click on a button to activate the keyboard navigation and then move around using WASD keys.

When you say "button" are you talking about the tool icon that starts the LookAndMoveTool, or clicking in the view?

It should not be necessary to click in the view to start navigation. If the LookAndMoveTool is the active view tool, pressing a navigation key should start navigation and then you only need to move the mouse without clicking to look around.

This is handled by ViewLookAndMove.enableKeyStart which is checked from ViewLookAndMove.onKeyTransition. If onKeyTransition isn't being called it could be an issue of where you have focus...
